CRAMv3.tex

```diff
@@ -2035,22 +2302,31 @@ \subsubsection*{Parameters}
 \textbf{Data type} & \textbf{Name} & \textbf{Comment}
 \tabularnewline
 \hline
-itf8[ ] & alphabet & list of all encoded values\tabularnewline
+itf8[ ] & alphabet & list of all encoded symbols (values)\tabularnewline
 \hline
 itf8[ ] & bit-lengths & array of bit-lengths for each symbol in the alphabet\tabularnewline
 \hline
 \end{tabular}
 
-\subsubsection*{Byte array coding}
+\subsubsection{Byte array coding}
+
+Often there is a need to encode an array of bytes where the length is not predetermined.
+For example the read identifiers differ per alignment record, possibly with different lengths, and this length must be stored somewhere.
+There are two choices available: storing the length explicitly (BYTE\_ARRAY\_LEN) or continuing to read bytes until a termination value is seen (BYTE\_ARRAY\_STOP).
 
-Often there is a need to encode an array of bytes. This can be optimized if the 
-length of the encoded arrays is known. For such cases BYTE\_ARRAY\_LEN and BYTE\_ARRAY\_STOP 
-codings can be used. 
+Note in contrast to this, quality values are known to be the same length as the sequence which is an already known quantity, so this does not need to be encoded using the byte array codecs.
 
-\subsubsection*{BYTE\_ARRAY\_LEN }
+\subsubsection*{BYTE\_ARRAY\_LEN: codec ID 4}
+
+Can encode types \textit{Byte[]}.
+
+Byte arrays are captured length-first, meaning that the length of every array element is written using an additional encoding.
+For example this could be a HUFFMAN encoding or another EXTERNAL block.
+The length is decoded first followed by the data, followed by the next length and data, and so on.
+
+This encoding can therefore be considered as a nested encoding, with each pair of nested encodings containing their own set of parameters.
+The byte stream for parameters of the BYTE\_ARRAY\_LEN encoding is therefore the concatenation of the length and value encoding parameters as described in section~\ref{subsec:writing-bytes}.
```

Comment:
Could you expand on this a little further with an example?  I'll make an attempt here:

As an example, if we used BETA encoding for length and EXTERNAL encoding for values, the parameters would be [ 6 (BETA encoding ID), len(beta params), beta params, 1 (EXTERNAL encoding ID), len(external params), external params ]

Reply:
Good idea.

I went with HUFFMAN + external.  This has the benefit of showing how constant sized values can still utilise the byte array len mechanism, but in an efficient manner that doesn't actually use any storage for length field.  (I use this for integer aux tags.)